toolchain/semantics/semantics_ir.cpp

```diff
@@ -206,9 +254,17 @@ auto SemanticsIR::StringifyType(SemanticsTypeId type_id) -> std::string {
     SemanticsNodeId node_id;
     // The index into node_id to print. Not used by all types.
     int index = 0;
+    // Whether this step is printed in a context that will be coerced to `type`.
+    bool type_context = true;
```

Comment:
🤷 

Error context offers a bit of "this is a type" hint, and `const` is also a type hint, so maybe the two hints together would generally be sufficient. It may even be that the error context is a sufficient hint and `as type` is more likely to confuse than help.

I mean, consider something like:

```
Cannot implicitly convert from `i32` to `()`.
```

In code, people write `fn F() -> ()`, not `fn F() -> () as type`. Maybe we don't need `as type` in error messages either?

Could also adjust error messages to do small things like "from" -> "from type" to hammer it in a little, without "as type" on the actual types.

Reply:
We should probably have two modes here, a "printing a value, make sure it's unambiguous" mode, and a "known based on context to be printing a type" mode. But I'll leave it as-is for now, with the `as type` printing moved to the end as suggested, and we can consider what to do here in later PRs.